File: frontend/src/objects/submission/JudgeFailureReason.ts
/** 判题失败原因枚举；只描述系统/构建/运行失败，不表示普通评测 verdict。对象对齐例外：该类型镜像 judge-protocol，不对应后端 submission domain object。 */
export type JudgeFailureReason =
  | 'judge_task_build_failed'
  | 'judger_runtime_failed'
  | 'checker_compile_failed'
  | 'checker_runtime_failed'
  | 'interactor_compile_failed'
  | 'interactor_runtime_failed'
  | 'problem_data_load_failed'
  | 'system_error'

const supportedJudgeFailureReasons = [
  'judge_task_build_failed',
  'judger_runtime_failed',
  'checker_compile_failed',
  'checker_runtime_failed',
  'interactor_compile_failed',
  'interactor_runtime_failed',
  'problem_data_load_failed',
  'system_error',
] as const satisfies readonly JudgeFailureReason[]

/** 判断字符串是否为受支持的判题失败原因；用于 API 响应防御性窄化。 */
export function isJudgeFailureReason(value: string): value is JudgeFailureReason {
  /** 注意：includes 需要把待测字符串临时断言为枚举联合类型，不会改变运行时值。 */
  return supportedJudgeFailureReasons.includes(value as JudgeFailureReason)
}

File: judge-protocol/src/main/scala/judgeprotocol/objects/response/JudgeFailureReason.scala
package judgeprotocol.objects.response

import io.circe.{Decoder, Encoder}

enum JudgeFailureReason:
  case JudgeTaskBuildFailed
  case JudgerRuntimeFailed
  case CheckerCompileFailed
  case CheckerRuntimeFailed
  case InteractorCompileFailed
  case InteractorRuntimeFailed
  case ProblemDataLoadFailed
  case SystemError

object JudgeFailureReason:
  def render(value: JudgeFailureReason): String =
    value match
      case JudgeFailureReason.JudgeTaskBuildFailed => "judge_task_build_failed"
      case JudgeFailureReason.JudgerRuntimeFailed => "judger_runtime_failed"
      case JudgeFailureReason.CheckerCompileFailed => "checker_compile_failed"
      case JudgeFailureReason.CheckerRuntimeFailed => "checker_runtime_failed"
      case JudgeFailureReason.InteractorCompileFailed => "interactor_compile_failed"
      case JudgeFailureReason.InteractorRuntimeFailed => "interactor_runtime_failed"
      case JudgeFailureReason.ProblemDataLoadFailed => "problem_data_load_failed"
      case JudgeFailureReason.SystemError => "system_error"

  given Encoder[JudgeFailureReason] = Encoder.encodeString.contramap(render)
  given Decoder[JudgeFailureReason] = Decoder.decodeString.emap {
    case "judge_task_build_failed" => Right(JudgeFailureReason.JudgeTaskBuildFailed)
    case "judger_runtime_failed" => Right(JudgeFailureReason.JudgerRuntimeFailed)
    case "checker_compile_failed" => Right(JudgeFailureReason.CheckerCompileFailed)
    case "checker_runtime_failed" => Right(JudgeFailureReason.CheckerRuntimeFailed)
    case "interactor_compile_failed" => Right(JudgeFailureReason.InteractorCompileFailed)
    case "interactor_runtime_failed" => Right(JudgeFailureReason.InteractorRuntimeFailed)
    case "problem_data_load_failed" => Right(JudgeFailureReason.ProblemDataLoadFailed)
    case "system_error" => Right(JudgeFailureReason.SystemError)
    case other => Left(s"Unsupported judge failure reason: $other")
  }
